mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala

```diff
@@ -534,8 +577,15 @@ class Word2VecModel private[spark] (
     // Need not divide with the norm of the given vector since it is constant.
     val cosVec = cosineVec.map(_.toDouble)
     var ind = 0
+    var vecNorm = 1f
+    if (norm) {
```

Comment:
You only need to normalize the top K values that are returned, at the end.
If speed is important here, then this implementation should probably be changed to use a heap or something rather than needlessly sort the whole thing. That's a much bigger bottleneck.